nucleus/payara-modules/requesttracing-core/src/main/java/fish/payara/nucleus/requesttracing/store/strategy/LongestTraceStorageStrategy.java

```diff
@@ -40,62 +40,40 @@
 package fish.payara.nucleus.requesttracing.store.strategy;
 
 import fish.payara.notification.requesttracing.RequestTrace;
-import java.util.ArrayList;
 import java.util.Collection;
-import java.util.Collections;
-import java.util.List;
+import java.util.Iterator;
 
 /**
- * Strategy for finding traces that need removing from a list. Will remove the
- * trace at the end of the
- * {@link RequestTrace RequestTrace} object's
- * natural ordering.
+ * Strategy for finding traces that need removing from a list. Will remove the trace at the shortest traces in the given
+ * set of traces. If multiple traces are equally short the oldest is removed first.
  */
 public class LongestTraceStorageStrategy implements TraceStorageStrategy {
 
     /**
-     * Gets the trace that needs removing. Sorts by the request trace's natural
-     * ordering.
+     * Gets the trace that needs removing. Removes the provided trace if present, or selects the shortest of the given
+     * traces should their number be larger then maxSize.
      *
-     * @param traces the list of traces to test.
-     * @param maxSize the maximum size of the list.
-     * @return the trace that needs removing, or null if no traces need
-     * removing.
+     * @see TraceStorageStrategy#getTraceForRemoval(Collection, int, RequestTrace)
      */
     @Override
-    public RequestTrace getTraceForRemoval(Collection<RequestTrace> traces, int maxSize) {
+    public RequestTrace getTraceForRemoval(Collection<RequestTrace> traces, int maxSize, RequestTrace traceToRemove) {
         if (traces == null || traces.isEmpty() || traces.size() <= maxSize) {
             return null;
         }
-        // Sort and return shortest request
-        List<RequestTrace> tracesCopy = new ArrayList<>(traces);
-        Collections.sort(tracesCopy);
-        return traces.toArray(new RequestTrace[]{})[traces.size() - 1];
```

Comment:
While this works this was just too painful performance wise to let it be - we do not know how large users might make the store - the target trace can be found by iterating instead of copying, sorting it (which copies again) and once again copy it to an array.